vision.md

```diff
@@ -206,26 +206,188 @@ The framework includes modules for cloud connectivity, data mapping, device
+      in a highly secured environment, the user is not required to access some
+      online documentation ressource, but is able to access all configuration
+      documentation via the binary shipped or derived from the thin-edge.io
+      framework
+    - Configurations are representable in multiple ways, e.g. as text or in a
+      more visual format (graph, image overview)
+1. The user os able to enable and disable functionality of the framework
+    - Not all functionalities of the framework are required in every use-case,
+      users are able to disable or even remove those parts from deployed
+      binaries
+    - For example: if only approved and audited components are allowed to be
+      used
+1. Deriving specialized implementations/binaries from the framework is possible
+    - The thin-edge.io ships a potentially large all-in-one binary with as much
+      functionality as the project provides, but
+    - Tailoring down the binary and removing unused functionality is easily
+      doable with minimal effort by a System Integrator
+    - Building a specialized binary with additional custom functionality is
+      easily doable by an Plugin Developer
+1. As little overhead as possible
+    1. in CPU time: The Framework ensures that the absolute minimum of CPU time
+       is consumed to deliver the desired functionality the user configured the
+       framework to do
+    1. in Memory usage: The Framework ensures that as little as possible memory
+       is in use, at every point in execution time, as possible
+1. Compatibility
+    - Configuration and components are written in such a way that they may be
+      changed in the future with no or minimal impact on their intended purpose
+
+### Functional requirements
+
+<!--
+The following list of functional requirements should be considered
+non-final. It might be extended in the future.
+-->
+
+1. The framework provides MQTT connectivity through a component
+1. All errors must be handled in a non-crashing way
+    - Unrecoverable errors may still cause the binary to shutdown eventually,
+      but not unexpectedly.
+1. The core implementation is written in Rust
+    - Plugins that connect to a thin-edge.io binary, may be written in another
+      programming language
+1. A deployment of thin-edge.io consists of a single binary with a single
+   configuration entry-point
+    - The final configuration may consists of more than one resource (file,
+      network resource, etc), and potentially be even loaded over the network,
+      this is left open
+1. The configuration is the single point of truth w.r.t. the initial state of
+   the components mentioned within
+    1. The default state of a plugin must be documented
+    1. A component may use a (documented) default value for a missing
+       configuration entry
+1. The configuration is the single point of truth for the communication between
+   components inside a single thin-edge.io binary deployment
```

Comment:
This is not clear to me. Do you mean that the communication  patterns must be defined in the configuration and cannot be pre-cabled in the executable? Why such a strong requirement?

Reply:
What use to have things setup statically and exposed at the same time? If things are hardcoded together then they are a single component. Otherwise they have to expose their inputs/outputs like everything else in the system.

The strong requirement is to make sure that this is something that can be:

1. Abstracted upon (things like interception, monitoring, as well as arbitrary inputs/outputs)
2. Can potentially be replaced for something else if the user wishes, this way the developers do not get to force the user how a certain component is wired up!
3. This way Component developers are forced to expose and thus make available their input/outputs simplifying the ecosystem, and its reasoning (How do I send messages of type X around? Well, you just gotta find something that outputs them!)